### PR TITLE
Added new functionality to support `withArgs($closure)`to validate a list of arguments at once

### PR DESCRIPTION
**Pull request by beni0888**
This PR adds the new functionality requested in #459 . In this way, from now on it is possible to pass a closure to `withArgs()` method in order to validate a list of arguments at once. All the arguments passed in the call to the expected method will be passed to the closure, and if it evaluates to true, then the list of
arguments is assumed to have matched the expectation.

### CODE REVIEW

File: library/Mockery/Expectation.php [hunk 1]
```diff
@@ -331,7 +340,7 @@ protected function _matchArg($expected, &$actual)
         if ($expected instanceof \Mockery\Matcher\MatcherAbstract) {
             return $expected->match($actual);
         }
-        if (is_a($expected, '\Hamcrest\Matcher') || is_a($expected, '\Hamcrest_Matcher')) {
+        if ($expected instanceof \Hamcrest\Matcher || $expected instanceof \Hamcrest_Matcher) {
```
Comment (aik099) [inline library/Mockery/Expectation.php]: The `is_a` was used, because we're unsure if mentioned class is present. Probably now we can use `instanceof` because Hamcrest is a required dependency, but this isn't the topic of this PR.

Comment (beni0888) [inline library/Mockery/Expectation.php]: I'm sorry but I'm afraid that I don't see your point. I don't know what's the difference between using `is_a()` and using `instanceof` about this issue. If the target class wasn't present, `instanceof`would just return `false`, just like `is_a()` does, so I can't see the problem here. 

I agree with you about this isn't the topic of this PR, but I like to apply the boy scout rule when it is possible and I thought this was a good chance. Anyway, if there is any problem with that I can leave that commit out of this PR.

Comment (aik099) [inline library/Mockery/Expectation.php]: I like `boy scout rule`, but it actually hurts code reviews when used with combination of PR sending IMO, because it introduces changes not related to PR topic.

You can send another PR with just these changes (`is_a` change to `instanceof`).

Comment (davedevelopment) [inline library/Mockery/Expectation.php]: I think we get the idea, don't worry about issuing another PR, just consider it next time.

Comment (beni0888) [inline library/Mockery/Expectation.php]: @aik099 :+1:  you're right, it makes code reviews more difficult. I'll consider it next time :)

Comment (aik099) [inline library/Mockery/Expectation.php]: Thanks @beni0888 .

File: library/Mockery/Matcher/MultiArgumentClosure.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     * Actual passed by reference to preserve reference trail (where applicable)
+     * back to the original method parameter.
+     *
+     * @param mixed $actual
+     * @return bool
+     */
+    public function match(&$actual)
+    {
+        $closure = $this->_expected;
+        $requiredParamsNumber = (new \ReflectionFunction($closure))->getNumberOfRequiredParameters();
```
Comment (davedevelopment) [inline library/Mockery/Matcher/MultiArgumentClosure.php]: I'm not sure if we need this level of strictness here and we can avoid having to worry about trailing args etc

Comment (davedevelopment) [inline library/Mockery/Matcher/MultiArgumentClosure.php]: (Very open to discussion on it)

Comment (beni0888) [inline library/Mockery/Matcher/MultiArgumentClosure.php]: Yeah, I think you're right. I fact, it doesn't make any sense doing that verification here. In addition, it introduces a bug as it returns false if the number of received params is not exactly the same as the number of required params of the closure, so if the closure receives some of the optional params it's going to return false even if the param is valid. I'm going to fix it immediately. I feel so ashamed right now...

Comment (davedevelopment) [inline library/Mockery/Matcher/MultiArgumentClosure.php]: > I feel so ashamed right now...

Can't tell if serious or not, so just in case, please don't feel any shame!!! This is an awesome contribution and this is what PRs are for

Comment (aik099) [inline library/Mockery/Matcher/MultiArgumentClosure.php]: Agree with @davedevelopment opinion.

**Comment by davedevelopment** [comment]
@beni0888 this looks great, could you do one last thing and add a quick one line summary to the `CHANGELOG.md` file please, under the 1.0.0 heading, then I'll get this merged.

Thanks

File: library/Mockery/Matcher/MultiArgumentClosure.php (hunk 2 above)
Comment (beni0888) [inline library/Mockery/Matcher/MultiArgumentClosure.php]: Thanks guys! I was just being a little bit dramatic :-). But I really felt a little bit stupid for not detect that mistake before...

**Comment by beni0888** [comment]
@davedevelopment Of course, I'll add it immediately. I forgot that little detail

**Comment by beni0888** [comment]
@davedevelopment It's done :)

**Comment by davedevelopment** [comment]
Thanks!
